prov/util/src/util_mem_monitor.c

```diff
@@ -34,96 +35,282 @@
 #include <ofi_mr.h>
 
 
-void ofi_monitor_init(struct ofi_mem_monitor *monitor)
+static struct ofi_uffd uffd;
+struct ofi_mem_monitor *uffd_monitor = &uffd.monitor;
+
+
+/*
+ * Initialize all available memory monitors
+ */
+void ofi_monitor_init(void)
+{
+	fastlock_init(&uffd_monitor->lock);
+	dlist_init(&uffd_monitor->list);
+}
+
+void ofi_monitor_cleanup(void)
 {
-	ofi_atomic_initialize32(&monitor->refcnt, 0);
+	assert(dlist_empty(&uffd_monitor->list));
 }
 
-void ofi_monitor_cleanup(struct ofi_mem_monitor *monitor)
+int ofi_monitor_add_cache(struct ofi_mem_monitor *monitor,
+			  struct ofi_mr_cache *cache)
 {
-	assert(ofi_atomic_get32(&monitor->refcnt) == 0);
+	int ret;
+
+	fastlock_acquire(&monitor->lock);
+	if (dlist_empty(&monitor->list) && (monitor == uffd_monitor))
+		ret = ofi_uffd_init();
+	else
+		ret = 0;
```

Comment:
What does it mean when monitor->list is empty and monitor is not uffd_monitor? Does this allow for different monitors to be added in the future and a no-op today?

Reply:
Yes, a different monitor in the future would need to update this check and call its own init function.  This would be better written as:
```
if (dlist_empty(...)) {
    if (monitor == uffd_monitor)
        ret = ofi_uffd_init();
    else
        ret = -FI_ENOSYS;
```

Reply:
Returning an error seems right.

Reply:
See my comment on ofi_monitor_init. Doesn't seem like it's designed for other monitors to exist. Also putting this as a comment here would help someone know how to add a new monitor.